drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,438 @@
+#define AXP192_LDO2_VOLTAGE_STEP_UV    (100000u)
+#define AXP192_LDO2_ENABLE_REG_MASK    (0x04u)
+#define AXP192_LDO2_ENABLE_REG_VAL     (0x04u)
+#define AXP192_LDO2_CURRENT_MAX_UA     (200u)
+
+/* LDO3 parameters */
+#define AXP192_LDO3_VOLTAGE_REG_MASK   (0x0Fu)
+#define AXP192_LDO3_VOLTAGE_REG_BITPOS (0u)
+#define AXP192_LDO3_VOLTAGE_RANGE_MAX  (AXP192_LDO3_VOLTAGE_REG_MASK)
+#define AXP192_LDO3_VOLTAGE_RANGE_MIN  (0x00u)
+#define AXP192_LDO3_VOLTAGE_MIN_UV     (1800000u)
+#define AXP192_LDO3_VOLTAGE_MAX_UV     (3300000u)
+#define AXP192_LDO3_VOLTAGE_STEP_UV    (100000u)
+#define AXP192_LDO3_ENABLE_REG_MASK    (0x08u)
+#define AXP192_LDO3_ENABLE_REG_VAL     (0x08u)
+#define AXP192_LDO3_CURRENT_MAX_UA     (200u)
+
+struct regulator_axp192_desc {
+	const uint8_t enable_reg;
+	const uint8_t enable_mask;
+	const uint8_t enable_val;
+	const uint8_t vsel_reg;
+	const uint8_t vsel_mask;
+	const uint8_t vsel_bitpos;
+	const int32_t max_ua;
+	const uint8_t num_ranges;
+	const struct linear_range *ranges;
+};
+
+struct regulator_axp192_common_config {
+	struct i2c_dt_spec i2c;
+};
+
+struct regulator_axp192_data {
+	struct regulator_common_data data;
+};
+
+struct regulator_axp192_config {
+	struct regulator_common_config common;
+	const struct regulator_axp192_desc *desc;
+	const struct device *mfd;
+
+	LOG_INSTANCE_PTR_DECLARE(log);
+};
+
+static struct linear_range dcdc1_ranges[] = {
+	LINEAR_RANGE_INIT(0u, 0u, 0u, 0u),
+	LINEAR_RANGE_INIT(AXP192_DCDC1_VOLTAGE_MIN_UV, AXP192_DCDC1_VOLTAGE_STEP_UV,
+			  AXP192_DCDC1_VOLTAGE_RANGE_MIN, AXP192_DCDC1_VOLTAGE_RANGE_MAX),
+};
+
+static const struct regulator_axp192_desc dcdc1_desc = {
+	.enable_reg = AXP192_REG_DCDC123_LDO23_CONTROL,
+	.enable_mask = AXP192_DCDC1_ENABLE_REG_MASK,
+	.enable_val = AXP192_DCDC1_ENABLE_REG_VAL,
+	.vsel_reg = AXP192_REG_DCDC1_VOLTAGE,
+	.vsel_mask = AXP192_DCDC1_VOLTAGE_REG_MASK,
+	.vsel_bitpos = AXP192_DCDC1_VOLTAGE_REG_BITPOS,
+	.max_ua = AXP192_DCDC1_CURRENT_MAX_UA,
+	.ranges = dcdc1_ranges,
+	.num_ranges = ARRAY_SIZE(dcdc1_ranges),
+};
+
+static struct linear_range dcdc2_ranges[] = {
+	LINEAR_RANGE_INIT(0u, 0u, 0u, 0u),
+	LINEAR_RANGE_INIT(AXP192_DCDC2_VOLTAGE_MIN_UV, AXP192_DCDC2_VOLTAGE_STEP_UV,
+			  AXP192_DCDC2_VOLTAGE_RANGE_MIN, AXP192_DCDC2_VOLTAGE_RANGE_MAX),
+};
+
+static const struct regulator_axp192_desc dcdc2_desc = {
+	.enable_reg = AXP192_REG_DCDC123_LDO23_CONTROL,
+	.enable_mask = AXP192_DCDC2_ENABLE_REG_MASK,
+	.enable_val = AXP192_DCDC2_ENABLE_REG_VAL,
+	.vsel_reg = AXP192_REG_DCDC2_VOLTAGE,
+	.vsel_mask = AXP192_DCDC2_VOLTAGE_REG_MASK,
+	.vsel_bitpos = AXP192_DCDC2_VOLTAGE_REG_BITPOS,
+	.max_ua = AXP192_DCDC2_CURRENT_MAX_UA,
+	.ranges = dcdc2_ranges,
+	.num_ranges = ARRAY_SIZE(dcdc2_ranges),
+};
+
+static struct linear_range dcdc3_ranges[] = {
+	LINEAR_RANGE_INIT(0u, 0u, 0u, 0u),
+	LINEAR_RANGE_INIT(AXP192_DCDC3_VOLTAGE_MIN_UV, AXP192_DCDC3_VOLTAGE_STEP_UV,
+			  AXP192_DCDC3_VOLTAGE_RANGE_MIN, AXP192_DCDC3_VOLTAGE_RANGE_MAX),
```

Comment:
note: since linear ranges are _self-documented_, better place values here instead of defines

Reply:
That was something I struggled with as well. From the codeines in my company I am conditioned to strictly use constants for all magic numbers. I corrected this in the driver to improve readability.